vmupdate/vmupdate.py

```diff
@@ -90,17 +111,124 @@ def get_targets(args, app):
                 f"Unknown qube name{'s' if plural else ''}"
                 f": {', '.join(unknowns) if plural else ''.join(unknowns)}"
             )
+    else:
+        targets += smart_targeting(app, args)
 
-    # remove dom0 - not a target
-    targets = [vm for vm in targets if vm.name != 'dom0']
+    # remove skipped qubes and dom0 - not a target
+    to_skip = args.skip.split(',')
+    targets = [vm for vm in targets
+               if vm.name != 'dom0' and vm.name not in to_skip]
     return targets
 
 
-def run_update(qube_predicator, targets, args):
-    qubes_to_go = [vm for vm in targets if qube_predicator(vm.klass)]
-    runner = update_manager.UpdateManager(qubes_to_go, args)
+def smart_targeting(app, args):
+    targets = []
+    for vm in app.domains:
+        if getattr(vm, 'updateable', False) and vm.klass != 'AdminVM':
+            try:
+                to_update = vm.features.get('updates-available', False)
+            except qubesadmin.exc.QubesDaemonCommunicationError:
+                to_update = False
+
+            if not to_update:
+                to_update = stale_update_info(vm, args)
+
+            if to_update:
+                targets.append(vm)
+
+    return targets
+
+
+def stale_update_info(vm, args):
+    today = datetime.today()
+    try:
+        last_update_str = vm.features.get(
+            'last-updates-check',
+            datetime.fromtimestamp(0).strftime('%Y-%m-%d %H:%M:%S')
+        )
+        last_update = datetime.fromisoformat(last_update_str)
+        if (today - last_update).days > args.n:
+            return True
+    except qubesadmin.exc.QubesDaemonCommunicationError:
+        pass
+    return False
+
+def run_update(targets, args, qube_klass="qubes"):
+    if args.dry_run:
+        print(f"Following {qube_klass} will be updated:",
+              ",".join((target.name for target in targets)))
+        return 0
+
+    runner = update_manager.UpdateManager(targets, args)
     return runner.run(agent_args=args)
 
 
+def restart_app_vms(args, templates):
+    if not args.restart:
+        return
+
+    templates_to_shutdown = [template for template in templates
+                             if template.is_running()]
+    if args.dry_run:
+        print("Following templates will be shutdown:",
+              ",".join((target.name for target in templates_to_shutdown)))
+        # we do not check if any volume is outdated, we expect it will be.
+        to_restart = {vm
+                      for template in templates
+                      for vm in template.appvms
+                      if vm.klass in ('AppVM', 'DispVM')
+                      and vm.is_running()}
+        print("Following qubes CAN be restarted:",
+              ",".join((target.name for target in to_restart)))
+        return
+
+    # first shutdown templates to apply changes to the root volume
+    for template in templates_to_shutdown:
+        try:
+            template.shutdown()
```

Comment:
you need to wait for shutdown to complete before restarting appvms. Take a look at https://github.com/QubesOS/qubes-core-admin-client/blob/main/qubesadmin/events/utils.py#L42 for less busy-wait approach.